[CI] Tripping assertion in PeerFinder#closePeers

**Build scan:**
https://gradle-enterprise.elastic.co/s/bcxstpjolnb2q/tests/:x-pack:plugin:ccr:internalClusterTest/org.elasticsearch.xpack.ccr.FollowerFailOverIT/testReadRequestsReturnLatestMappingVersion

**Reproduction line:**
```
./gradlew ':x-pack:plugin:ccr:internalClusterTest' --tests "org.elasticsearch.xpack.ccr.FollowerFailOverIT.testReadRequestsReturnLatestMappingVersion" -Dtests.seed=A75C1F0C6F541957 -Dtests.locale=fr-CA -Dtests.timezone=America/Tijuana -Druntime.java=20
```

**Applicable branches:**
8.10

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.ccr.FollowerFailOverIT&tests.test=testReadRequestsReturnLatestMappingVersion

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=498, name=elasticsearch[follower2][clusterApplierService#updateTask][T#1], state=RUNNABLE, group=TGRP-FollowerFailOverIT]

  at __randomizedtesting.SeedInfo.seed([A75C1F0C6F541957:A564CB31A58082CC]:0)

  Caused by: java.lang.AssertionError: (No message provided)

    at __randomizedtesting.SeedInfo.seed([A75C1F0C6F541957]:0)
    at org.elasticsearch.discovery.PeerFinder.closePeers(PeerFinder.java:182)
    at org.elasticsearch.cluster.coordination.Coordinator.onClusterStateApplied(Coordinator.java:415)
    at org.elasticsearch.cluster.coordination.Coordinator.lambda$handleApplyCommit$8(Coordinator.java:403)
    at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.lambda$map$0(ActionListenerImplementations.java:108)
    at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:89)
    at org.elasticsearch.cluster.service.ClusterApplierService$ClusterApplyActionListener.onResponse(ClusterApplierService.java:598)
    at org.elasticsearch.cluster.service.ClusterApplierService$ClusterApplyActionListener.onResponse(ClusterApplierService.java:569)
    at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:439)
    at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:916)
    at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:217)
    at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:183)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
    at java.lang.Thread.run(Thread.java:1623)

```

--- a/docs/changelog/100986.yaml
+++ b/docs/changelog/100986.yaml
@@ -1,0 +1,6 @@
+pr: 100986
+summary: Synchronize Coordinator#onClusterStateApplied
+area: Cluster Coordination
+type: bug
+issues:
+ - 99023

--- a/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
+++ b/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
@@ -414,13 +414,15 @@ public class Coordinator extends AbstractLifecycleComponent implements ClusterSt
 
     private void onClusterStateApplied() {
         assert ThreadPool.assertCurrentThreadPool(ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME);
-        if (getMode() != Mode.CANDIDATE) {
-            joinHelper.onClusterStateApplied();
-            closeElectionScheduler();
-            peerFinder.closePeers();
-        }
-        if (getLocalNode().isMasterNode()) {
-            joinReasonService.onClusterStateApplied(applierState.nodes());
+        synchronized (mutex) {
+            if (getMode() != Mode.CANDIDATE) {
+                joinHelper.onClusterStateApplied();
+                closeElectionScheduler();
+                peerFinder.closePeers();
+            }
+            if (getLocalNode().isMasterNode()) {
+                joinReasonService.onClusterStateApplied(applierState.nodes());
+            }
         }
     }
 

--- a/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
+++ b/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
@@ -415,14 +415,14 @@ public class Coordinator extends AbstractLifecycleComponent implements ClusterSt
     private void onClusterStateApplied() {
         assert ThreadPool.assertCurrentThreadPool(ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME);
         synchronized (mutex) {
-            if (getMode() != Mode.CANDIDATE) {
+            if (mode != Mode.CANDIDATE) {
                 joinHelper.onClusterStateApplied();
                 closeElectionScheduler();
                 peerFinder.closePeers();
             }
-            if (getLocalNode().isMasterNode()) {
-                joinReasonService.onClusterStateApplied(applierState.nodes());
-            }
+        }
+        if (getLocalNode().isMasterNode()) {
+            joinReasonService.onClusterStateApplied(applierState.nodes());
         }
     }
 

